### PR TITLE
core/state/snapshot: implement disk iterator seek

### PR DESCRIPTION
**Pull request by holiman**
This PR changes the way the Iteratee interface is defined. Previously, we had
```
NewIteratorWithPrefix(prefix []byte)
NewIteratorWithStart(start []byte)
```
But no good way to get a prefixed iterator with a given start position. As that was missing, a lot of code made one of two mistakes: 

1. Use `NewIteratorWithStart( prefix ++ start)`. This solution is wrong; consider the prefix `0x0a` and start `0x00`. The returned iterator would also return e.g `0x0b`, as it is after `a` in the keyspace. 
2. Use `NewIteratorWithPrefix( prefix ++ start)`. This is also wrong. Consider the prefix `0x0a` and start `0x01`. This iterator would miss a key at `0x0a02`, since `0x0a01` is not a prefix to `0x0a02`. 

I replaced both of these with 
```
NewIteratorWith(prefix, start []byte)
```

### CODE REVIEW

**Comment by holiman** [comment]
This wound up to become pretty large, and it may affect a few regions where I'm not sure we have tests. 
For example `les/clientpool.go`: 
```golang

func (db *nodeDB) getPosBalanceIDs(start, stop enode.ID, maxCount int) (result []enode.ID) {
	if maxCount <= 0 {
		return
	}
	it := db.db.NewIteratorWithStart(db.key(start.Bytes(), false))
```
I'm not sure if the db here is already prefixed 'natively'? If so, that should be fine. 
For `trie/sync_bloom.go`: 
```golang

	for it.Next() && atomic.LoadUint32(&b.closed) == 0 {
		// If the database entry is a trie node, add it to the bloom
		if key := it.Key(); len(key) == common.HashLength {
			b.bloom.Add(syncBloomHasher(key))
			bloomLoadMeter.Mark(1)
		}
		// If enough time elapsed since the last iterator swap, restart
		if time.Since(swap) > 8*time.Second {
			key := common.CopyBytes(it.Key())

			it.Release()
			it = database.NewIteratorWithStart(key)
```
In this case, I'm fairly certain this will break, since `it.Key()` will return the `prefix + position`. If there's a 'native' prefix, it will now be applied twice


In general, I'm leaning towards removing the `NewIteratorWithStart` and `NewIteratorWithPrefix` in favour of `NewIteratorWith(prefix, start []byte)`, as it's semantically cleaner.

**Comment by karalabe** [comment]
This new API seems super strange to me. I would have expected that `start` should also include `prefix`, not for the two to be concatenated.

**Comment by karalabe** [comment]
> I replaced both of these with
> NewIteratorWith(prefix, start []byte)

That's a bad name. `NewIteratorWith("a", "b")`. You can't use the parameter name to extend the function name, that info gets lost in code. Is adding a 3rd method that bad? `NewIteratorWithPrefixAndStart`?

**Comment by holiman** [comment]
> This new API seems super strange to me. I would have expected that start should also include prefix, not for the two to be concatenated.

That is one way of doing it. But what if the input is `prefix:"a", start: ""`. How should that be interpreted? 
I think it's cleaner to interpret the args as

1. Within the set of keys prefixed by `prefix`, 
2. Start at key `prefix+start`

Can you formulate what the corresponding ruleset would be for the api-style that you suggest?

**Comment by holiman** [comment]
> Is adding a 3rd method that bad? NewIteratorWithPrefixAndStart?

I can do that, sure. I first changed all implementations, to ensure that all usages of it were correct, so that `NewIteratorWithPrefixAndStart(prefix, nil) == NewIteratorWithPrefix(prefix)` and  `NewIteratorWithPrefixAndStart(nil, start) == NewIteratorWithPrefix(start)`. 

And after having done that  change, I thought that the two extra methods there really wasn't motivating their own existence sufficiently.

**Comment by holiman** [comment]
TODO; change `NewIteratorWith` to `NewIterator`, remove old `NewIterator`

**Comment by holiman** [comment]
Done

File: core/state/snapshot/disklayer_test.go [hunk 1]
```diff
@@ -432,4 +435,81 @@ func TestDiskPartialMerge(t *testing.T) {
 // This test case is a tiny specialized case of TestDiskPartialMerge, which tests
 // some very specific cornercases that random tests won't ever trigger.
 func TestDiskMidAccountPartialMerge(t *testing.T) {
+	// TODO(@karalabe) ?
+}
+
+func tempDB() (ethdb.Database, error) {
```
Comment (karalabe) [inline core/state/snapshot/disklayer_test.go:441]: Nit, please inline this into the test, there's only one use for it, so might as well have one less dangling helper method. That will allow to `os.RemoveAll` the folder after you're done. It's super annoying when tests leave their files on disk after they are done.

File: ethdb/leveldb/leveldb.go [hunk 2]
```diff
@@ -488,3 +476,12 @@ func (r *replayer) Delete(key []byte) {
 	}
 	r.failure = r.writer.Delete(key)
 }
+
+// BytesPrefixRange returns key range that satisfy
+// - the given prefix, and
+// - the given seek position
+func BytesPrefixRange(prefix, start []byte) *util.Range {
```
Comment (karalabe) [inline ethdb/leveldb/leveldb.go:483]: Is there a reason to keep this public?

File: ethdb/memorydb/memorydb.go [hunk 3]
```diff
@@ -129,55 +129,28 @@ func (db *Database) NewBatch() ethdb.Batch {
...
-}
-
-// NewIteratorWithPrefix creates a binary-alphabetical iterator over a subset
-// of database content with a particular key prefix.
-func (db *Database) NewIteratorWithPrefix(prefix []byte) ethdb.Iterator {
+// NewIterator creates a binary-alphabetical iterator over a subset
+// of database content with a particular key prefix, starting at a particular
+// initial key (or after, if it does not exist).
+// Note: This method assumes that the prefix is NOT part of the start, so there's
+// no need for the caller to prepend the prefix to the start
```
Comment (karalabe) [inline ethdb/memorydb/memorydb.go:136]: Please move this `Note` into the interface definition. That's where it's more important. Also please leave an empty comment line above the `Note` section, keeps the thing cleaner.

**Comment by holiman** [comment]
Fixed

**Pull request merged**